mne/preprocessing/otp.py

```diff
@@ -0,0 +1,112 @@
+        Raw data to denoise.
+    duration : float
+        The window duration (in seconds) to use (default: 10.).
+    scalings : dict | None
+        If None (default), use ``dict(mag=1e15, grad=1e13, eeg=1e6)``.
+        These defaults will scale different channel types to approximately
+        at the same unit.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose)
+
+    Returns
+    -------
+    raw_clean : instance of Raw
+        The cleaned data.
+
+    Notes
+    -----
+    This algorithm is computationally expensive, and can be several times
+    slower than realtime for conventional M/EEG datasets.
+
+    This algorithm is covered by a provisional patent.
+
+    .. versionadded:: 0.14
+
+    References
+    ----------
+    .. [1] Taulu, S and Larson E (2016). "Reducing noise in
+           electromagnetic sensor arrays using oversampled temporal
+           projection", presented at the 2016 International Conference
+           on Biomagnetism, Seoul, Korea 3 Oct 2016.
+    """
+    # XXX eventually offer option to treat "bads" by projecting them onto
+    # others, but not use them in bases for others?
+    logger.info('Processing MEG data using oversampled temporal projection')
+    scalings = _check_scalings_user(scalings)
+    n_samp = int(round(float(duration) * raw.info['sfreq']))
+    del duration
+    if n_samp / raw.info['sfreq'] > raw.times[-1]:
+        raise ValueError('Effective duration (%s) must be at most the '
+                         'duration of the raw instance (%s)'
+                         % (n_samp / raw.info['sfreq'], raw.times[-1]))
+    raw = raw.copy().load_data()
+    picks = list(_pick_data_channels(raw.info, exclude='bads'))
+    read_lims = np.arange(0, len(raw.times) + 1, n_samp)  # include endpt
+    if read_lims[-1] - read_lims[-2] != n_samp:
+        read_lims[-1] = len(raw.times)
+    min_samp = np.diff(read_lims).min()
+    if (min_samp <= len(picks)).any():
+        raise ValueError('duration (%s) yielded %s samples, which is fewer '
+                         'than the number of channels (%s)'
+                         % (n_samp / raw.info['sfreq'], min_samp, len(picks)))
+    pl = 's' if len(read_lims) != 2 else ''
+    logger.info('    Processing %s data chunk%s of (at least) %0.1f sec'
+                % (len(read_lims) - 1, pl, n_samp / raw.info['sfreq']))
+    apply_picks_list = _picks_by_type(pick_info(raw.info, picks))
+    for start, stop in zip(read_lims[:-1], read_lims[1:]):
+        logger.info('    Denoising % 8.2f - % 8.2f sec'
+                    % tuple(raw.times[[start, stop - 1]]))
+        data_full = raw._data[:, start:stop]  # view (for correct operation)
+        data_picked = data_full[picks]  # not a view (for correct operation)
+        if not np.isfinite(data_picked).all():
+            raise RuntimeError('non-finite data (inf or nan) found in raw '
+                               'instance')
```

Comment:
this should be checked earlier, so that we don't need to start running the algorithm before we hit the error